File: kagent/python/packages/kagent-adk/src/kagent/adk/_approval.py
"""before_tool_callback implementation for HITL tool approval.

Uses the ADK-native ToolContext.request_confirmation() mechanism.
When a tool in the approval set is invoked:
  - First call: requests confirmation via tool_context, blocks execution.
  - Re-invocation after user responds: checks tool_context.tool_confirmation.
"""

import logging
from typing import Any

from google.adk.agents.callback_context import CallbackContext
from google.adk.models.llm_request import LlmRequest
from google.adk.tools.base_tool import BaseTool
from google.adk.tools.tool_context import ToolContext

logger = logging.getLogger(__name__)


def strip_confirmation_parts_callback(
    callback_context: CallbackContext,
    llm_request: LlmRequest,
) -> None:
    """Before-model callback that strips adk_request_confirmation parts from the LLM request.

    These are synthetic ADK HITL events the LLM never produced and does not need
    to reason about. The session still stores them for ADK's resume machinery.
    """
    if not llm_request.contents:
        return None
    filtered_contents = []
    for content in llm_request.contents:
        parts = [
            p
            for p in (content.parts or [])
            if not (p.function_call and p.function_call.name == "adk_request_confirmation")
            and not (p.function_response and p.function_response.name == "adk_request_confirmation")
        ]
        if parts:
            content.parts = parts
            filtered_contents.append(content)
    llm_request.contents = filtered_contents
    return None


def make_approval_callback(
    tools_requiring_approval: set[str],
):
    """Create a before_tool_callback that requests confirmation for specified tools.

    Args:
        tools_requiring_approval: Set of tool names that need human approval.

    Returns:
        A callback compatible with Google ADK's before_tool_callback signature.
    """

    def before_tool(
        tool: BaseTool,
        args: dict[str, Any],
        tool_context: ToolContext,
    ) -> str | dict | None:
        tool_name = tool.name
        if tool_name not in tools_requiring_approval:
            return None  # No approval needed, proceed normally

        # On re-invocation after confirmation, ADK populates tool_confirmation
        if tool_context.tool_confirmation is not None:
            if tool_context.tool_confirmation.confirmed:
                logger.debug("Tool %s approved by user, proceeding", tool_name)
                return None  # Approved — proceed with tool execution
            logger.debug("Tool %s rejected by user", tool_name)
            # Check for an optional rejection reason in the payload
            # (the key "rejection_reason" is set by _agent_executor._process_hitl_decision)
            payload = tool_context.tool_confirmation.payload or {}
            reason = payload.get("rejection_reason", "") if isinstance(payload, dict) else ""
            # __build_response_event wraps it as {"result": "..."} that LLM adapters expect
            # ADK will skip executing the function if the before tool callback returns a response
            if reason:
                return f"Tool call was rejected by user. Reason: {reason}"
            return "Tool call was rejected by user."

        # First invocation — request confirmation and block execution
        # This response is never sent to the LLM
        logger.debug("Tool %s requires approval, requesting confirmation", tool_name)
        tool_context.request_confirmation(
            hint=f"Tool '{tool_name}' requires approval before execution.",
        )
        return {"status": "confirmation_requested", "tool": tool_name}

    return before_tool

File: kagent/go/adk/pkg/agent/approval.go
package agent

import (
	"fmt"

	"google.golang.org/adk/agent"
	"google.golang.org/adk/agent/llmagent"
	adkmodel "google.golang.org/adk/model"
	"google.golang.org/adk/tool"
	"google.golang.org/genai"
)

// stripConfirmationPartsCallback is a BeforeModelCallback that removes
// adk_request_confirmation FunctionCall and FunctionResponse parts from the
// LLM request before it reaches any model provider. These are synthetic ADK
// HITL events the LLM never produced and does not need to reason about.
// The session still stores them so ADK's resume machinery can find them.
func MakeStripConfirmationPartsCallback() llmagent.BeforeModelCallback {
	return func(_ agent.CallbackContext, req *adkmodel.LLMRequest) (*adkmodel.LLMResponse, error) {
		out := make([]*genai.Content, 0, len(req.Contents))
		for _, c := range req.Contents {
			if c == nil {
				continue
			}
			filtered := make([]*genai.Part, 0, len(c.Parts))
			for _, p := range c.Parts {
				if p == nil {
					continue
				}
				if p.FunctionCall != nil && p.FunctionCall.Name == "adk_request_confirmation" {
					continue
				}
				if p.FunctionResponse != nil && p.FunctionResponse.Name == "adk_request_confirmation" {
					continue
				}
				filtered = append(filtered, p)
			}
			if len(filtered) == 0 {
				continue
			}
			newContent := &genai.Content{
				Role:  c.Role,
				Parts: filtered,
			}
			out = append(out, newContent)
		}
		req.Contents = out
		return nil, nil
	}
}

// MakeApprovalCallback creates a BeforeToolCallback that gates execution of
// tools in the approval set behind request_confirmation / ToolConfirmation.
// Port of kagent-adk/src/kagent/adk/_approval.py:make_approval_callback().
func MakeApprovalCallback(toolsRequiringApproval map[string]bool) llmagent.BeforeToolCallback {
	return func(ctx tool.Context, t tool.Tool, args map[string]any) (map[string]any, error) {
		toolName := t.Name()

		// No approval needed for this tool.
		if !toolsRequiringApproval[toolName] {
			return nil, nil
		}

		// On re-invocation after confirmation, ADK populates ToolConfirmation.
		if confirmation := ctx.ToolConfirmation(); confirmation != nil {
			if confirmation.Confirmed {
				// Approved — proceed with tool execution.
				return nil, nil
			}
			// Rejected — extract optional rejection reason from payload.
			payload, _ := confirmation.Payload.(map[string]any)
			reason, _ := payload["rejection_reason"].(string)
			if reason != "" {
				return map[string]any{
					"result": fmt.Sprintf("Tool call was rejected by user. Reason: %s", reason),
				}, nil
			}
			return map[string]any{
				"result": "Tool call was rejected by user.",
			}, nil
		}

		// First invocation — request confirmation and block execution.
		if err := ctx.RequestConfirmation(
			fmt.Sprintf("Tool '%s' requires approval before execution.", toolName),
			nil,
		); err != nil {
			return nil, fmt.Errorf("failed to request confirmation for tool %s: %w", toolName, err)
		}
		return map[string]any{
			"status": "confirmation_requested",
			"tool":   toolName,
		}, nil
	}
}
